File: GraalPyProject_backup_20260112_195403/PiCarGraalController/src/main/java/nathanielnarofsky/todolist/Lgpio.java
package nathanielnarofsky.todolist;

public class Lgpio {

    //this class is a wrapper for the lgpio functions used in the gpio zero library
    //this is needed because the lgpio library is not directly accessible from Java via Py4J
    //this class will be used to call the underlying c functions via the foreign function and memory api

    public static Object SET_PULL_NONE() {
        //placeholder for the actual implementation
        System.out.print("SET_PULL_NONE called");
        return null;
    }

    public static Object SET_PULL_UP() {
        //placeholder for the actual implementation
        System.out.print("SET_PULL_UP called");
        return null;
    }

    public static Object SET_PULL_DOWN() {
        //placeholder for the actual implementation
        System.out.print("SET_PULL_DOWN called");
        return null;
    }


    public static Object SET_BIAS_DISABLE() {
        //placeholder for the actual implementation
        System.out.print("SET_BIAS_DISABLE called");
        return null;
    }

    public static Object SET_BIAS_PULL_UP() {
        //placeholder for the actual implementation
        System.out.print("SET_BIAS_PULL_UP called");
        return null;
    }

    public static Object SET_BIAS_PULL_DOWN() {
        //placeholder for the actual implementation
        System.out.print("SET_BIAS_PULL_DOWN called");
        return null;
    }

    // Edge detection constants
    public static Object RISING_EDGE() {
        //placeholder for the actual implementation
        System.out.print("RISING_EDGE called");
        return 1;
    }

    public static Object FALLING_EDGE() {
        //placeholder for the actual implementation
        System.out.print("FALLING_EDGE called");
        return 2;
    }

    public static Object BOTH_EDGES() {
        //placeholder for the actual implementation
        System.out.print("BOTH_EDGES called");
        return 3;
    }

    // Chip management functions
    public static Object gpiochip_open(int chip) {
        //placeholder for the actual implementation
        System.out.print("gpiochip_open called with chip: " + chip);
        return -1; 
    }

    public static Object gpiochip_close(int handle) {
        //placeholder for the actual implementation
        System.out.print("gpiochip_close called with handle: " + handle);
        return 0; 
    }

    // GPIO pin operations
    public static Object gpio_claim_input(int handle, int pin, Object pull) {
        //placeholder for the actual implementation
        System.out.print("gpio_claim_input called with handle: " + handle + ", pin: " + pin + ", pull: " + pull);
        return 0; 
    }

    public static Object gpio_claim_output(int handle, int pin, boolean value) {
        //placeholder for the actual implementation
        System.out.print("gpio_claim_output called with handle: " + handle + ", pin: " + pin + ", value: " + value);
        return 0; 
    }

    public static Object gpio_read(int handle, int pin) {
        //placeholder for the actual implementation
        System.out.print("gpio_read called with handle: " + handle + ", pin: " + pin);
        return false; 
    }

    public static Object gpio_write(int handle, int pin, boolean value) {
        //placeholder for the actual implementation
        System.out.print("gpio_write called with handle: " + handle + ", pin: " + pin + ", value: " + value);
        return 0; 
    }

    public static Object gpio_get_mode(int handle, int pin) {
        //placeholder for the actual implementation
        System.out.print("gpio_get_mode called with handle: " + handle + ", pin: " + pin);
        return 0; 
    }

    public static Object gpio_free(int handle, int pin) {
        //placeholder for the actual implementation
        System.out.print("gpio_free called with handle: " + handle + ", pin: " + pin);
        return 0; 
    }

    // PWM functions
    public static Object tx_pwm(int handle, int pin, double frequency, double dutyCycle) {
        //placeholder for the actual implementation
        System.out.print("tx_pwm called with handle: " + handle + ", pin: " + pin + ", freq: " + frequency + ", duty: " + dutyCycle);
        return 0; 
    }

    public static Object tx_wave(int handle, int pin, Object waveform) {
        //placeholder for the actual implementation
        System.out.print("tx_wave called with handle: " + handle + ", pin: " + pin);
        return 0; 
    }

    // Event/callback functions (simplified signatures)
    public static Object gpio_start_alerts(int handle, int pin, int edges) {
        //placeholder for the actual implementation
        System.out.print("gpio_start_alerts called with handle: " + handle + ", pin: " + pin + ", edges: " + edges);
        return 0; 
    }

    public static Object gpio_stop_alerts(int handle, int pin) {
        //placeholder for the actual implementation
        System.out.print("gpio_stop_alerts called with handle: " + handle + ", pin: " + pin);
        return 0; 
    }

}

File: GraalPyProject/PiCarGraalController/src/main/resources/_lgpio.py
"""
_lgpio module implementation for GraalPy

This module provides a Python interface to the lgpio library by delegating
calls to the Java LgpioNativeBridge class. This allows the existing lgpio_extra.py
wrapper to work unchanged with GraalPy.

The module exposes all the lgpio functions with the exact same names and signatures
that the original C-based _lgpio module provides.
"""

import java

# Import the Java bridge class
try:
    LgpioBridge = java.type('nathanielnarofsky.todolist.LgpioNativeBridge')
    print("✅ Successfully imported LgpioNativeBridge")
    
    # Check if native library loaded successfully
    if not LgpioBridge.isLibraryLoaded():
        error = LgpioBridge.getLibraryError()
        raise RuntimeError(f"Native lgpio library not loaded: {error}")
        
except Exception as e:
    print(f"❌ Failed to import LgpioNativeBridge: {e}")
    raise

# GPIO Chip Operations
def _gpiochip_open(gpiochip):
    """Opens a gpiochip device"""
    return LgpioBridge._gpiochip_open(int(gpiochip))

def _gpiochip_close(handle):
    """Closes a gpiochip device"""
    return LgpioBridge._gpiochip_close(int(handle))

# GPIO Info Operations  
def _gpio_get_chip_info(handle):
    """Gets information about a gpiochip"""
    result = LgpioBridge._gpio_get_chip_info(int(handle))
    # Convert Java object to expected format if needed
    return result

def _gpio_get_line_info(handle, gpio):
    """Gets information about a gpiochip line"""
    result = LgpioBridge._gpio_get_line_info(int(handle), int(gpio))
    return result

def _gpio_get_mode(handle, gpio):
    """Gets the mode of a GPIO"""
    return LgpioBridge._gpio_get_mode(int(handle), int(gpio))

# GPIO Claim Operations
def _gpio_claim_input(handle, lFlags, gpio):
    """Claims a GPIO for input"""
    return LgpioBridge._gpio_claim_input(int(handle), int(lFlags), int(gpio))

def _gpio_claim_output(handle, lFlags, gpio, level):
    """Claims a GPIO for output"""
    return LgpioBridge._gpio_claim_output(int(handle), int(lFlags), int(gpio), int(level))

def _gpio_claim_alert(handle, lFlags, eFlags, gpio, notify_handle):
    """Claims a GPIO for alerts"""
    return LgpioBridge._gpio_claim_alert(int(handle), int(lFlags), int(eFlags), int(gpio), int(notify_handle))

def _gpio_free(handle, gpio):
    """Frees a GPIO"""
    return LgpioBridge._gpio_free(int(handle), int(gpio))

# GPIO Read/Write Operations
def _gpio_read(handle, gpio):
    """Reads the value of a GPIO"""
    return LgpioBridge._gpio_read(int(handle), int(gpio))

def _gpio_write(handle, gpio, level):
    """Writes a value to a GPIO"""
    return LgpioBridge._gpio_write(int(handle), int(gpio), int(level))

# Group Operations (simplified versions for basic functionality)
def _group_claim_input(handle, lFlags, gpios):
    """Claims a group of GPIO for inputs"""
    # This would need more complex conversion for the gpios parameter
    # For now, return a placeholder - implement as needed
    return 0

def _group_claim_output(handle, lFlags, gpios, levels):
    """Claims a group of GPIO for outputs"""
    # This would need more complex conversion for the gpios/levels parameters
    # For now, return a placeholder - implement as needed  
    return 0

def _group_free(handle, gpio):
    """Frees a group of GPIO"""
    return LgpioBridge._group_free(int(handle), int(gpio))

def _group_read(handle, gpio):
    """Reads a group of GPIO"""
    result = LgpioBridge._group_read(int(handle), int(gpio))
    return result

def _group_write(handle, gpio, group_bits, group_mask):
    """Writes to a group of GPIO"""
    return LgpioBridge._group_write(int(handle), int(gpio), int(group_bits), int(group_mask))

# PWM/Pulse Operations
def _tx_pulse(handle, gpio, pulse_on, pulse_off, pulse_offset, pulse_cycles):
    """Starts pulses on a GPIO"""
    return LgpioBridge._tx_pulse(int(handle), int(gpio), int(pulse_on), 
                                 int(pulse_off), int(pulse_offset), int(pulse_cycles))

def _tx_pwm(handle, gpio, pwm_frequency, pwm_duty_cycle, pulse_offset, pulse_cycles):
    """Starts PWM on a GPIO"""
    return LgpioBridge._tx_pwm(int(handle), int(gpio), float(pwm_frequency), 
                               float(pwm_duty_cycle), int(pulse_offset), int(pulse_cycles))

def _tx_servo(handle, gpio, pulse_width, servo_frequency, pulse_offset, pulse_cycles):
    """Starts servo pulses on a GPIO"""
    return LgpioBridge._tx_servo(int(handle), int(gpio), int(pulse_width), 
                                 int(servo_frequency), int(pulse_offset), int(pulse_cycles))

def _tx_wave(handle, gpio, pulses):
    """Starts a wave on a group of GPIO"""
    # This would need conversion of the pulses parameter
    # For now, return a placeholder - implement as needed
    return 0

def _tx_busy(handle, gpio, kind):
    """Checks if transmission is active on a GPIO"""
    return LgpioBridge._tx_busy(int(handle), int(gpio), int(kind))

def _tx_room(handle, gpio, kind):
    """Checks transmission queue space on a GPIO"""
    return LgpioBridge._tx_room(int(handle), int(gpio), int(kind))

# GPIO Configuration
def _gpio_set_debounce_micros(handle, gpio, debounce_micros):
    """Sets the debounce time for a GPIO"""
    return LgpioBridge._gpio_set_debounce_micros(int(handle), int(gpio), int(debounce_micros))

def _gpio_set_watchdog_micros(handle, gpio, watchdog_micros):
    """Sets the watchdog time for a GPIO"""
    return LgpioBridge._gpio_set_watchdog_micros(int(handle), int(gpio), int(watchdog_micros))

# I2C Operations
def _i2c_open(i2c_bus, i2c_address, i2c_flags):
    """Opens an I2C device"""
    return LgpioBridge._i2c_open(int(i2c_bus), int(i2c_address), int(i2c_flags))

def _i2c_close(handle):
    """Closes an I2C device"""
    return LgpioBridge._i2c_close(int(handle))

def _i2c_write_quick(handle, bit):
    """SMBus write quick"""
    return LgpioBridge._i2c_write_quick(int(handle), int(bit))

def _i2c_read_byte(handle):
    """SMBus read byte"""
    return LgpioBridge._i2c_read_byte(int(handle))

def _i2c_write_byte(handle, byte_val):
    """SMBus write byte"""
    return LgpioBridge._i2c_write_byte(int(handle), int(byte_val))

def _i2c_read_byte_data(handle, reg):
    """SMBus read byte data"""
    return LgpioBridge._i2c_read_byte_data(int(handle), int(reg))

def _i2c_write_byte_data(handle, reg, byte_val):
    """SMBus write byte data"""
    return LgpioBridge._i2c_write_byte_data(int(handle), int(reg), int(byte_val))

def _i2c_read_word_data(handle, reg):
    """SMBus read word data"""
    return LgpioBridge._i2c_read_word_data(int(handle), int(reg))

def _i2c_WriteWordData(handle, reg, word_val):
    """SMBus write word data"""
    return LgpioBridge._i2c_WriteWordData(int(handle), int(reg), int(word_val))

def _i2c_process_call(handle, reg, word_val):
    """SMBus process call"""
    return LgpioBridge._i2c_process_call(int(handle), int(reg), int(word_val))

# I2C Block Operations (simplified - would need proper byte array conversion)
def _i2c_read_block_data(handle, reg):
    """SMBus read block data"""
    # This would need proper conversion from Java byte arrays to Python
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _i2c_write_block_data(handle, reg, data):
    """SMBus write block data"""
    # This would need proper conversion from Python to Java byte arrays
    # For now, return a placeholder - implement as needed
    return 0

def _i2c_block_process_call(handle, reg, data):
    """SMBus block process call"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _i2c_write_i2c_block_data(handle, reg, data):
    """Writes I2C block data"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return 0

def _i2c_read_i2c_block_data(handle, reg, count):
    """Reads I2C block data"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _i2c_read_device(handle, count):
    """Reads from raw I2C device"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _i2c_write_device(handle, data):
    """Writes to raw I2C device"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return 0

def _i2c_zip(handle, data):
    """Performs multiple I2C transactions"""
    # This would need complex conversion
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

# Notification Operations
def _notify_open():
    """Opens a notification handle"""
    return LgpioBridge._notify_open()

def _notify_close(handle):
    """Closes a notification handle"""
    return LgpioBridge._notify_close(int(handle))

def _notify_pause(handle):
    """Pauses notifications"""
    return LgpioBridge._notify_pause(int(handle))

def _notify_resume(handle):
    """Resumes notifications"""
    return LgpioBridge._notify_resume(int(handle))

# Serial Operations
def _serial_open(tty, baud, ser_flags):
    """Opens a serial device"""
    return LgpioBridge._serial_open(str(tty), int(baud), int(ser_flags))

def _serial_close(handle):
    """Closes a serial device"""
    return LgpioBridge._serial_close(int(handle))

def _serial_read_byte(handle):
    """Reads a byte from serial device"""
    return LgpioBridge._serial_read_byte(int(handle))

def _serial_write_byte(handle, byte_val):
    """Writes a byte to serial device"""
    return LgpioBridge._serial_write_byte(int(handle), int(byte_val))

def _serial_read(handle, count):
    """Reads bytes from serial device"""
    # This would need proper conversion from Java to Python bytes
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _serial_write(handle, data):
    """Writes bytes to serial device"""
    # This would need proper conversion from Python to Java bytes
    # For now, return a placeholder - implement as needed
    return 0

def _serial_data_available(handle):
    """Returns number of bytes available for reading"""
    return LgpioBridge._serial_data_available(int(handle))

# SPI Operations
def _spi_open(spi_device, spi_channel, baud, spi_flags):
    """Opens an SPI device"""
    return LgpioBridge._spi_open(int(spi_device), int(spi_channel), int(baud), int(spi_flags))

def _spi_close(handle):
    """Closes an SPI device"""
    return LgpioBridge._spi_close(int(handle))

def _spi_read(handle, count):
    """Reads bytes from SPI device"""
    # This would need proper conversion from Java to Python bytes
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

def _spi_write(handle, data):
    """Writes bytes to SPI device"""
    # This would need proper conversion from Python to Java bytes
    # For now, return a placeholder - implement as needed
    return 0

def _spi_xfer(handle, data):
    """Transfers bytes with SPI device"""
    # This would need proper conversion
    # For now, return a placeholder - implement as needed
    return [0, bytearray()]

# Module initialization
print("🔗 _lgpio bridge module loaded successfully")
print(f"📚 Native library status: {'✅ Loaded' if LgpioBridge.isLibraryLoaded() else '❌ Not loaded'}")